[Improved Search] Show predictive suggestions

sztomek opened the pull request:
## Description
This PR displays the search suggestions on the search screen.
Analytics and the recently announced language code unification is not part of the scope.

Designs: 9aiIqQo7ibm7R1AyIYr8fH-fi-1_1574
P2: pdeCcb-aqh-p2


Fixes https://linear.app/a8c/issue/PCDROID-170/update-search-screen

## Testing Instructions
1. Clean install the app
2. Press 'Get started' on onboarding, try filtering some podcasts on the recommendations step
- [ ] search still works there
3. Once you're done with onboarding, tap 'Discover'
4. Tap the search bar
5. Start typing
- [ ] suggestions appear as you type
6. Select a suggestion
- [ ] real search kicks in and you'll see podcasts and episodes
- [ ] tapped term will be saved as history item
7. On the suggestions, tap the '+' icon on any podcast suggestion
- [ ] verify that the podcasts followed status updates after a short period of time

Please also verify:
- Search functionality remains intact with the FF turned OFF
- Search on podcasts tab still works with/without the FF

## Screenshots or Screencast 

https://github.com/user-attachments/assets/314ee3f8-8370-4670-8fbd-286276f9ef8f



## Checklist
- [x] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [ ] ~I have considered whether it makes sense to add tests for my changes~
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
- [ ] ~I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.~
 
#### I have tested any UI changes...
- [x] with different themes
- [x] with a landscape orientation
- [x] with the device set to have a large display and font size
- [x] for accessibility with TalkBack

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

MiSikora reviewed (commented):
Apart from using `hashCode()` as a key, I don't there are any issues with the code itself. However, I found a few UI and UX issues worth addressing:

### Incorrect highlight range

Currently, the implementation highlights the first `N` characters of the word but the search term can appear anywhere within the result.

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/9983df99-279d-414c-a506-ba14d61133c7" />

### Small follow tap target

The follow button is quite small, making it easy to misclick.

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/261e3c0b-897c-4b7c-9170-4f7b4325769e" />

### Button highlight doesn’t extend edge to edge

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/bcfdfb73-59d6-481e-aec0-5f3b80bde30c" />

### Follow state is not responsive

There’s about a 2-second delay between tapping the follow button and seeing the followed state.

https://github.com/user-attachments/assets/f250759c-7c8e-4f0e-978a-acf8400d884d

sztomek commented on modules/features/search/src/main/java/au/com/shiftyjelly/pocketcasts/search/SearchAutoCompleteResultsPage.kt:
```diff
@@ -0,0 +1,286 @@
+            contentPadding = PaddingValues(bottom = bottomInset),
+        ) {
+            results.forEachIndexed { index, item ->
+                item(key = item.hashCode(), contentType = "content-${item.javaClass}") {
```
hi
> ❓ From what I understand about the feature it isn't a dynamic list that gets manipulated. We load it and display it. Does the key help here?

you're right. removed the keys.

> ❓ How many items do we expect here to be present? Right now it shows 4 suggestions and 4 results. Is it expected to go into dozens or more? If it is then hashCode() isn't a good key candidate as hash collision can lead to crashes. If it is not then key isn't really important since we're displaying only a couple of items.

4 items are currently what we expect to display from each section. removed the key anyways.

sztomek commented:
thanks for your review @MiSikora 
I've addressed all your concerns in 77ea7d9b3a34b4112017fd3a8071361d93c7cf0b
also answered your questions, i believe.
Pls take another look 👀

MiSikora commented on modules/features/search/src/main/java/au/com/shiftyjelly/pocketcasts/search/SearchAutoCompleteResultsPage.kt:
```diff
@@ -0,0 +1,297 @@
+                color = MaterialTheme.theme.colors.primaryText02,
+            )
+        }
+        if (item.isSubscribed) {
+            Icon(
+                modifier = Modifier
+                    .padding(start = 8.dp)
+                    .size(32.dp),
+                painter = painterResource(IR.drawable.ic_check),
+                contentDescription = null,
+                tint = MaterialTheme.theme.colors.support02,
+            )
+        } else {
+            Icon(
+                painter = painterResource(IR.drawable.ic_add_black_24dp),
+                contentDescription = null,
+                tint = MaterialTheme.theme.colors.secondaryText02,
+                modifier = Modifier
+                    .padding(start = 8.dp)
+                    .size(32.dp)
+                    .clickable(onClick = onFollow),
+            )
+        }
```
ℹ️ Now the icons feel a bit too large.

<img width="1080" height="423" alt="Image" src="https://github.com/user-attachments/assets/b4a54ba5-2a5e-438e-879f-76360d52ae89" />

I'd suggest sticking to the standard sizing provided by the Material Design with `minimumInteractiveComponentSize()` and `IconButton`. Something like: 

```kotlin
if (item.isSubscribed) {
    Icon(
        modifier = Modifier.minimumInteractiveComponentSize(),
        painter = painterResource(IR.drawable.ic_check),
        contentDescription = null,
        tint = MaterialTheme.theme.colors.support02,
    )
} else {
    IconButton(onClick = onFollow) {
        Icon(
            painter = painterResource(IR.drawable.ic_add_black_24dp),
            contentDescription = null,
            tint = MaterialTheme.theme.colors.secondaryText02,
        )
    }
}
```

This way we can get expected touch targets and a circle touch indicator.

<img width="1080" height="1452" alt="Image" src="https://github.com/user-attachments/assets/de13689b-3f5b-4b75-a383-18d3a159b43e" />

MiSikora commented on modules/features/search/src/main/java/au/com/shiftyjelly/pocketcasts/search/SearchHandler.kt:
```diff
@@ -101,50 +106,80 @@ class SearchHandler @Inject constructor(
-                }.subscribeOn(Schedulers.io())
-                    .map<AutoCompleteSearch> {
-                        AutoCompleteSearch.Success(
+                flowOf(autoCompleteManager.autoCompleteSearch(term = query))
```
🚫 I suggested `flowOf()` only for the empty case because this can cause a crash now. Turn on the airplane mode and type in anything. It is because exception happens not while emitting but before a `Flow` is even created. This should be either changed back to `flow{}` or `autoCompleteManager.autoCompleteSearch(term = query)` should be surrounded with a `try/catch`.
